posts/2017-06-02-devops-serverless-variables.md

```diff
@@ -0,0 +1,155 @@
+
+In this article we’ll explore some ideas on how you can use these features to better automate your serverless operations.
+
+## Dyamic Configuration
+You can use the variable system to dynamically generate configuration data for your config files with the help of JS scripts, which is normally not supported by YAML/JSON. Some examples include generating random numbers, dates, doing mathematical computation, or fetching remote data. Let’s look at two of these examples..
+
+
+### Generating Random ID
+Let’s say you want to create a bucket with a unique id attached to the name. It wouldn’t be possible to generate that random id inside of serverless.yml without the magic of programming. You can utilize the variable system for this use case like this:
+
+```yml
+# serverless.yml
+
+Resources:
+  Resources:
+    myBucket:
+      Type: AWS::S3::Bucket
+      Properties:
+        BucketName: my-bucket-with-id-${file(vars.js):bucketId}
+```
+
+And in vars.js
+```js
+module.exports.bucketId = (new Date()).getTime().toString();
+```
+
+
+
+
+### Fetching Remote Data
+You can go even further with JS scripts by working with promises and async operations, which is supported out of the box by the framework. That means you can fetch data needed for your configuration from any remote source you want.
+
+For example, if you need to use your Github first name as the function name, you can fetch that data from the Github API using a vars.js file like this:
+
+```
+module.exports.githubFirstName = () => {
+  /*
+   * some async logic here calling the Github API...
+   */
+  const githubFirstName = “Sam”;
+  return Promise.resolve(githubFirstName);
+};
+```
+Then in serverless.yml you can do something like this:
+
+```
+functions:
+  myPersonalFunction:
+    name: ${file(vars.js):githubFirstName}
+```
+The use cases of this feature are limitless! You can fetch global config from DynamoDB, get sensitive data from KMS, or interact with your own API. Whatever is needed by your application.
+
+## Predeployment Processing
+A super creative usage of the variable system is to use it without having any actual variables at all! One use case of that is doing some pre-deployment processing as a build script, maybe to prepare or build certain AWS resources that your service depends on or make some changes to your config files. This works out of the box because processing variables happens at the beginning on Serverless initialization. You’ll just need to write your script and reference it wherever needed, most likely in the ‘custom` property of your `serverless.yml`.
```

Comment:
The backtick before `custom` is the wrong way, which is messing up the formatting.